src/Builder.hs

```diff
@@ -111,22 +121,25 @@ instance NFData   Builder
 -- 'Stage' and GHC 'Package').
 builderProvenance :: Builder -> Maybe Context
 builderProvenance = \case
-    DeriveConstants  -> context Stage0 deriveConstants
-    GenApply         -> context Stage0 genapply
-    GenPrimopCode    -> context Stage0 genprimopcode
+    DeriveConstants  -> context Stage1 deriveConstants
```

Comment:
`Stage0` to `Stage1`? What does this imply?

Reply:
The general theme is that what is build in `stageN` is placed in `stage(N+1)` (this does hold for all binaries, but not yet for the package.db, which I'll need to fix.

The result is that every binary is built in `stageN` is in `stage(N+1)/bin` and any library built in `stageN` in `stage(N+1)/lib`, as a result building the binary distribution for stageN is simply zipping the `stageN/{bin, lib}` folders.

For some tools, which are used only during the build, where we know we do not need to rebuild them with the stage1 compiler, and the ones built with the bootstrap (stage0) compiler are good, we can fix the location.

So the stage0 -> stage1 changes are mostly due to the changed layout of where binaries end up.

Reply:
> The general theme is that what is build in `stageN` is placed in `stage(N+1)`

@angerman But this is not how Hadrian does things now. The `_build/stage0` folder contains all build artefacts build by Stage0 compiler (i.e. by the bootstrapping compiler), including binaries.

I think this is a simple and logical arrangement. Why do you want to place binaries built in Stage0 into `_build/stage1`?

Reply:
In particular, I think the Stage1 GHC binary should go somewhere into `_build/stage0/ghc` directory, because it's built during Stage0. 

Similarly, the Stage2 GHC should go somewhere into `_build/stage1/ghc` directory, because it's built during Stage1. 

Or is it confusing for some reason?

Reply:
@snowleopard true. `stage0` contains the objects and with the current (not this PR) the products. This however poses a problem when trying to separate the stages cleanly.

Stage0 is our bootstrap stage. Our bootstrap compiler is our stage0 compiler. As such that one should be `_build/stage1/bin/ghc` IMO. Similarly the stage1 compiler built by the stage0 compiler should subsequently be `_build/stage1/bin/ghc`.

The other reason is, if I'm talking about packaging up stageN, I would basically zip `_ build/stageN/{bin,lib}`. And if the stage1 compiler would live in stage0 this wouldn't work anymore. Unless I put bin into stage0, but then I zip `_build/stage0/{bin,lib}` to build the binary distribution of the stage1 compiler, which is kinda confusing I think.

If we separate the notion of build artifacts and products (binaries and installed libraries), then the build artifacts (objects) are in the stage they are build in, but the "installation" of the binaries and libraries happens to stage+1.

Reply:
> This however poses a problem when trying to separate the stages cleanly.

@angerman What are these problems, specifically? I still don't understand.

> And if the stage1 compiler would live in stage0 this wouldn't work anymore.

Why? Just zip `_build/stage0/{bin,lib}` -- what's wrong with that?

> which is kinda confusing I think

Is everything boils down to the fact that `_build/stage0/bin/ghc-stage1.exe` is confusing? Is that the only 'problem' you are referring to above, or there anything else?

Let's look at this from the documentation perspective.

At the moment we can say: "Everything build in Stage0 is in `_build/stage0`", "Everything build in Stage1 is in `_build/stage1`".

With your approach we have to say: "Everything build in Stage0 is in `_build/stage0` plus `_build/stage1/{bin,lib}`", "Everything build in Stage1 is in `_build/stage1`, minus `_build/stage1/{bin,lib}` plus `_build/stage2/{bin,lib}`".

Now which one is more confusing?

And this doesn't stop in the documentation only. This complexity propagates to the build system itself. For example, to freeze Stage1 we no longer can simply ask Shake to skip rebuilding everything in `_build/stage1`. We will have to exclude `_build/stage1/{bin,lib}` and include `_build/stage2/{bin,lib}`. 

Is all this complexity needed just so that you have `_build/stage1/bin/ghc-stage1.exe` instead of `_build/stage0/bin/ghc-stage1.exe`? I find it a too high price to pay just for the sake of matching 0 and 1 in this path.